control_panel_api/authentication.py

```diff
@@ -1,37 +1,70 @@
+import logging
+
+import requests
 from django.conf import settings
-import jwt
-from jwt.exceptions import InvalidTokenError
+from jose import jwt
+from jose.exceptions import JWTError
+from requests.exceptions import RequestException
 from rest_framework.authentication import (
     BaseAuthentication,
-    get_authorization_header
+    get_authorization_header,
 )
 from rest_framework.exceptions import AuthenticationFailed
 
 from control_panel_api.models import User
 
+logger = logging.getLogger(__name__)
 
-class Auth0JWTAuthentication(BaseAuthentication):
 
-    def authenticate(self, request):
+def get_jwks():
+    response = requests.get(
+        f'https://{settings.OIDC_DOMAIN}/.well-known/jwks.json')
+    return response.json()
+
+
+def get_key(jwks, unverified_header):
+    for jwk in jwks.get('keys'):
+        if jwk['kid'] == unverified_header['kid']:
+            return jwk
 
+    return None
+
+
+class Auth0JWTAuthentication(BaseAuthentication):
+    def authenticate(self, request):
```

Comment:
Should this method be broken up into a few smaller methods? There's quite a lot going on here now

Reply:
I realise it looks big, but if you look each method call is doing something specific and then the errors are clearly handled in this controlling function block (which I would rather have than some error handling here and some in other functions).

Reply:
I think as well it would be good if we split this in smaller, well-named functions if possible.

Reply:
I agree, the code in `authenticate` is too low level - it could be split out to look something like this:
```python
def authenticate(self, request):
    jwt = self.get_jwt(request)  # throws AuthenticationFailed if JWT couldn't be found/decoded
    return self.get_or_create_user(jwt.sub)
```

Reply:
Just realised that this is what @xoen said earlier